torch/csrc/jit/tensorexpr/cuda_codegen.cpp

```diff
@@ -301,43 +264,98 @@ static bool CheckEqual(const Expr* lhs, const Expr* rhs) {
   return immediateEquals(diff_s.node(), 0);
 }
 
-// Identify the pattern: a[e1] = a[e1] + e2.
-static bool isAtomicAdd(const Store* v, const Expr** atomic_add_value) {
-  ScalarType dtype = v->value()->dtype().scalar_type();
-  if (dtype != ScalarType::Float && dtype != ScalarType::Double) {
-    return false;
-  }
-  const Add* add_v = dynamic_cast<const Add*>(v->value());
-  if (!add_v) {
-    return false;
-  }
-  const Load* load_v = dynamic_cast<const Load*>(add_v->lhs());
-  if (!load_v) {
-    return false;
-  }
-  if (v->base_handle() != load_v->base_handle()) {
-    return false;
-  }
-  if (v->indices().empty() && load_v->indices().empty()) {
-    return false;
-  }
-  bool index_equal = CheckEqual(v->flat_index(), load_v->flat_index());
-  if (index_equal) {
-    *atomic_add_value = add_v->rhs();
+class AtomicAddFuser : public IRMutator {
+ public:
+  AtomicAddFuser(
+      const std::unordered_set<const Var*>& thread_local_bufs,
+      const GPUMetaVarRewriter* metavars)
```

Comment:
Any reason why "metavars" is not a const ref, similar to "thread_local_bufs"? The most common reason for a const pointer is that it could be nullptr, but it doesn't seem to be the case here.

Reply:
The only reason was to match the local style of the file: storing as a unique_ptr and passing as ptr (see CudaAnalysis and it's use by CudaPrinter).

NBD to change this one, but wouldn't want to add a general cuda codgen cleanup into this PR.